# Formation 2022  

##  Initiation à l’analyse de données Minion pour l'analyse de métagénome viraux

Created by J. Orjuela (DIADE-IRD), D. Filloux (PHIM-CIRAD) and A. Comte (PHIM-IRD) and from a formation by J. Orjuela (DIADE-IRD), F. Sabot (DIADE-IRD) and G. Sarah (AGAP-INRAE) given in Septembre 2021

Septembre 2022

# TP4 : De novo assembly of the Metavirome

Metagenome assembly is the process of transforming a set of short, overlapping, and potentially erroneous DNA segments from environmental samples into the accurate representation of the underlying microbiomes's genomes. Metagenome assembly is an efficient approach to deciphering the “microbial dark matter” in the microbiota based on metagenomic sequencing.

The objective is to reconstruct viruses in presence in the dataset.

A lot of assembler tools exists for long reads. Here we will focus on FLYE which is fast and often really accurate:

- Flye : https://github.com/fenderglass/Flye

But you can also have a look at SPAdes, an other assembler working with metagenome.

- SPAdes : https://github.com/ablab/spades

## 1. Assemble with Flye:

Flye is a de novo assembler for single-molecule sequencing reads, such as those produced by PacBio and Oxford Nanopore Technologies. It is designed for a wide range of datasets, from small bacterial projects to large mammalian-scale assemblies. The package represents a complete pipeline: it takes raw PacBio / ONT reads as input and outputs polished contigs. Flye also has a special mode for metagenome assembly.

In [5]:
# create working repository
mkdir -p ~/work/SG-ONT-2022/ASSEMBLY/FLYE
cd ~/work/SG-ONT-2022/ASSEMBLY/FLYE

### 1.1. Launch Flye:

In [3]:
flye --help

usage: flye (--pacbio-raw | --pacbio-corr | --pacbio-hifi | --nano-raw |
	     --nano-corr | --nano-hq ) file1 [file_2 ...]
	     --out-dir PATH

	     [--genome-size SIZE] [--threads int] [--iterations int]
	     [--meta] [--polish-target] [--min-overlap SIZE]
	     [--keep-haplotypes] [--debug] [--version] [--help] 
	     [--scaffold] [--resume] [--resume-from] [--stop-after] 
	     [--read-error float] [--extra-params]

Assembly of long reads with repeat graphs

optional arguments:
  -h, --help            show this help message and exit
  --pacbio-raw path [path ...]
                        PacBio regular CLR reads (<20% error)
  --pacbio-corr path [path ...]
                        PacBio reads that were corrected with other methods
                        (<3% error)
  --pacbio-hifi path [path ...]
                        PacBio HiFi reads (<1% error)
  --nano-raw path [path ...]
                        ONT regular reads, pre-Guppy5 (<20% error)
  --nano-corr path [path ...]
     

In [5]:
# this can take time to run
flye --meta --nano-hq ../../CLEANING/reads_vs_ananas_unmapped.fastq -o out_flye

[2022-07-29 08:07:41] INFO: Starting Flye 2.9-b1768
[2022-07-29 08:07:41] INFO: >>>STAGE: configure
[2022-07-29 08:07:41] INFO: Configuring run
[2022-07-29 08:07:43] INFO: Total read length: 173940017
[2022-07-29 08:07:43] INFO: Reads N50/N90: 435 / 276
[2022-07-29 08:07:43] INFO: Minimum overlap set to 1000
[2022-07-29 08:07:43] INFO: >>>STAGE: assembly
[2022-07-29 08:07:43] INFO: Assembling disjointigs
[2022-07-29 08:07:43] INFO: Reading sequences
[2022-07-29 08:07:45] INFO: Building minimizer index
[2022-07-29 08:07:45] INFO: Pre-calculating index storage
0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 
[2022-07-29 08:07:49] INFO: Filling index
0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 
[2022-07-29 08:07:58] INFO: Extending reads
[2022-07-29 08:08:03] INFO: Overlap-based coverage: 22
[2022-07-29 08:08:03] INFO: Median overlap divergence: 0.0844806
0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 
[2022-07-29 08:08:39] INFO: Assembled 54 disjointigs
[2022-07-29 08:08:39] INFO: Generating se

**How many contig do we have after the assembly?**

### 1.2. Estimate the quality of the assembly:

#### 1.2.1. QUAST

QUAST evaluates genome assemblies.

QUAST works both with and without a reference genome.
The tool accepts multiple assemblies, thus is suitable for comparison. 

http://quast.sourceforge.net/quast

In [15]:
metaquast.py -h

MetaQUAST: Quality Assessment Tool for Metagenome Assemblies
Version: 5.2.0

Usage: python /opt/conda/envs/quast/bin/metaquast.py [options] <files_with_contigs>

Options:
-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]
-r   <filename,filename,...>      Comma-separated list of reference genomes or directory with reference genomes
--references-list <filename>      Text file with list of reference genome names for downloading from NCBI
-g  --features [type:]<filename>  File with genomic feature coordinates in the references (GFF, BED, NCBI or TXT)
                                  Optional 'type' can be specified for extracting only a specific feature type from GFF
-m  --min-contig  <int>           Lower threshold for contig length [default: 500]
-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]

Advanced options:
-s  --split-scaffolds                 Split assemblies by continuous fragments

In [19]:
metaquast.py ../../ASSEMBLY/FLYE/out_flye/assembly.fasta --silent

/opt/conda/envs/quast/bin/metaquast.py ../../ASSEMBLY/FLYE/out_flye/assembly.fasta --silent


System information:
  OS: Linux-5.15.0-41-generic-x86_64-with-debian-bullseye-sid (linux_64)
  Python version: 3.7.12
  CPUs number: 8

Started: 2022-07-29 09:03:39

Logging to /home/jovyan/work/SG-ONT-2022/ASSEMBLY/FLYE/quast_results/results_2022_07_29_09_03_39/metaquast.log
NOTICE: Maximum number of threads is set to 2 (use --threads option to set it manually)

Contigs:
  Pre-processing...
  ../../ASSEMBLY/FLYE/out_flye/assembly.fasta ==> assembly

No references are provided, starting to search for reference genomes in SILVA 16S rRNA database and to download them from NCBI...

2022-07-29 09:03:40
NOTICE: Permission denied accessing /opt/conda/envs/quast/lib/python3.7/site-packages/quast_libs/silva. Silva will be downloaded to home directory /home/jovyan/.quast

ERROR! Failed downloading SILVA 16S rRNA gene database (http://www.arb-silva.de/fileadmin/silva_databases/release_138.1/Exports/SILV

**Observe QUAST output**

In [20]:
head -25 quast_results/latest/report.txt

All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., "# contigs (>= 0 bp)" and "Total length (>= 0 bp)" include all contigs).

Assembly                    assembly
# contigs (>= 0 bp)         32      
# contigs (>= 1000 bp)      32      
# contigs (>= 5000 bp)      2       
# contigs (>= 10000 bp)     0       
# contigs (>= 25000 bp)     0       
# contigs (>= 50000 bp)     0       
Total length (>= 0 bp)      88391   
Total length (>= 1000 bp)   88391   
Total length (>= 5000 bp)   12760   
Total length (>= 10000 bp)  0       
Total length (>= 25000 bp)  0       
Total length (>= 50000 bp)  0       
# contigs                   32      
Largest contig              6877    
Total length                88391   
GC (%)                      44.93   
N50                         2612    
N90                         1952    
auN                         3266.0  
L50                         12      
L90                         27      
# N's per 100 kbp          

#### 1.2.2. CheckV

CheckV assesses the quality and completeness of metagenome-assembled viral genomes

In [15]:
checkv --help

CheckV v0.9.0: assessing the quality of metagenome-assembled viral genomes
https://bitbucket.org/berkeleylab/checkv

usage: checkv <program> [options]

programs:
    end_to_end          run full pipeline to estimate completeness, contamination, and identify closed genomes
    contamination       identify and remove host contamination on integrated proviruses
    completeness        estimate completeness for genome fragments
    complete_genomes    identify complete genomes based on terminal repeats and flanking host regions
    quality_summary     summarize results across modules
    download_database   download the latest version of CheckV's database
    update_database     update CheckV's database with your own complete genomes

optional arguments:
  -h, --help  show this help message and exit


In [16]:
# download database
checkv download_database ./


CheckV v0.9.0: download_database
[1/4] Checking latest version of CheckV's database...
[2/4] Downloading 'checkv-db-v1.2'...
[3/4] Extracting 'checkv-db-v1.2'...
[4/4] Building DIAMOND database...
Run time: 531.07 seconds
Peak mem: 1.32 GB


In [22]:
export CHECKVDB=~/work/SG-ONT-2022/ASSEMBLY/FLYE/checkv-db-v1.2

In [23]:
checkv end_to_end out_flye/assembly.fasta output_checkv


CheckV v0.9.0: contamination
[1/8] Reading database info...
[2/8] Reading genome info...
[3/8] Calling genes with Prodigal...
[4/8] Reading gene info...
[5/8] Running hmmsearch...
[6/8] Annotating genes...
[7/8] Identifying host regions...
[8/8] Writing results...
Run time: 65.22 seconds
Peak mem: 0.22 GB

CheckV v0.9.0: completeness
[1/8] Skipping gene calling...
[2/8] Initializing queries and database...
[3/8] Running DIAMOND blastp search...
[4/8] Computing AAI...
[5/8] Running AAI based completeness estimation...
[6/8] Running HMM based completeness estimation...
[7/8] Determining genome copy number...
[8/8] Writing results...
Run time: 25.22 seconds
Peak mem: 1.36 GB

CheckV v0.9.0: complete_genomes
[1/7] Reading input sequences...
[2/7] Finding complete proviruses...
[3/7] Finding direct/inverted terminal repeats...
[4/7] Filtering terminal repeats...
[5/7] Checking genome for completeness...
[6/7] Checking genome for large duplications...
[7/7] Writing results...
Run time: 0.0 

**observe the different output files**

Is there any interresting high-quality viral contig?

Contig_31 seems to be of high quality, viral and has a size of 6877 nucleotides.

What specie / genus of virus is this contig?

### 1.3. Taxonomic assignation of contigs

In [26]:
# create working repository
mkdir -p ~/work/SG-ONT-2022/ASSEMBLY/FLYE/DIAMOND
cd ~/work/SG-ONT-2022/ASSEMBLY/FLYE/DIAMOND

We are going to reuse the viral diamond database used in TP3

In [27]:
#create a symbolic link of the diamond database
ln -s ~/work/SG-ONT-2022/ASSIGNATION/DIAMOND/viral.dmnd viral.dmnd

In [29]:
diamond blastx --quiet -d viral.dmnd --outfmt 6 stitle qtitle pident length mismatch gapopen qstart qend sstart send evalue bitscore -q ../out_flye/assembly.fasta -o diamond-matches.csv

In [34]:
grep "contig_31" diamond-matches.csv

YP_004935358.1 ORF1 gene product [Actinidia virus B]	contig_31	37.8	1697	852	34	6728	2184	4	1678	0.0	997
YP_009551967.1 Replicase [Grapevine virus J]	contig_31	36.6	1711	855	36	6728	2184	4	1680	4.11e-318	974
YP_009551905.1 putative replicase [Grapevine virus H]	contig_31	37.0	1704	866	38	6728	2181	4	1688	1.27e-316	970
YP_009664756.1 putative replicase, partial [Actinidia virus A]	contig_31	37.4	1702	828	37	6665	2184	14	1686	1.24e-315	967
NP_619654.1 putative replicase [Grapevine virus B]	contig_31	37.5	1694	862	30	6728	2181	4	1679	2.94e-314	964
YP_006590065.1 replicase [Grapevine virus F]	contig_31	35.8	1713	886	36	6728	2184	4	1701	1.69e-303	936
YP_009389463.1 RNA-dependent RNA polymerase [Grapevine virus K]	contig_31	35.5	1700	885	38	6728	2184	4	1676	8.41e-303	933
NP_619662.1 putative replicase [Grapevine virus A]	contig_31	35.6	1696	897	32	6728	2181	4	1684	1.65e-299	925
YP_009373228.1 RNA-dependent RNA polymerase [Agave tequilana leaf virus]	contig_31	38.0	1564	874	34	6713	2181	9	152

Looks like contig_31 is a Vitivirus.

### 1.4. Visualise the coverage of the reads on contig 31

#### 1.4.1. Remapping of the reads on contig31

In [38]:
mkdir -p ~/work/SG-ONT-2022/ASSEMBLY/FLYE/CONTIG31
cd ~/work/SG-ONT-2022/ASSEMBLY/FLYE/CONTIG31

In [39]:
##extract contig31 from the multifasta file
samtools faidx ../out_flye/assembly.fasta contig_31 > contig_31.fasta

In [44]:
minimap2 -ax map-ont contig_31.fasta ../../../CLEANING/reads_vs_ananas_unmapped.fastq > contig31_vs_reads.sam

[M::mm_idx_gen::0.001*2.25] collected minimizers
[M::mm_idx_gen::0.002*2.34] sorted minimizers
[M::main::0.002*2.31] loaded/built the index for 1 target sequence(s)
[M::mm_mapopt_update::0.003*2.13] mid_occ = 48
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1
[M::mm_idx_stat::0.003*2.00] distinct minimizers: 1302 (99.92% are singletons); average occurrences: 1.035; average spacing: 5.102
[M::worker_pipeline::182.124*2.98] mapped 405689 sequences
[M::main] Version: 2.17-r941
[M::main] CMD: minimap2 -ax map-ont contig_31.fasta ../../../CLEANING/reads_vs_ananas_unmapped.fastq
[M::main] Real time: 182.147 sec; CPU: 543.586 sec; Peak RSS: 0.588 GB


In [45]:
samtools flagstats contig31_vs_reads.sam

729154 + 0 in total (QC-passed reads + QC-failed reads)
268755 + 0 secondary
54710 + 0 supplementary
0 + 0 duplicates
392535 + 0 mapped (53.83% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


In [46]:
samtools view -b -S contig31_vs_reads.sam > contig31_vs_reads.bam

In [47]:
samtools sort contig31_vs_reads.bam -o contig31_vs_reads_sorted.bam

In [48]:
samtools index contig31_vs_reads_sorted.bam

In [49]:
samtools coverage contig31_vs_reads_sorted.bam

#rname	startpos	endpos	numreads	covbases	coverage	meandepth	meanbaseq	meanmapq
contig_31	1	6877	123780	6877	100	2070.44	19.9	12.8


#### 1.4.2. Visualise the coverage / depht of the reads on the contig on TABLET

- 1. Open tablet
- 2. click on Open Assembly
- 3. import the sorted bam as primary assembly and the contig_31.fasta as reference.

Explore the data